core/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java

```diff
@@ -172,6 +184,44 @@ public Result applyFailedShards(ClusterState clusterState, List<FailedRerouteAll
         return buildResultAndLogHealthChange(allocation, "shards failed [" + failedShardsAsString + "] ...");
     }
 
+    private ClusterState removeStaleAllocationIds(ClusterState clusterState, List<FailedRerouteAllocation.StaleShard> staleShards) {
+        MetaData oldMetaData = clusterState.metaData();
+        MetaData.Builder metaDataBuilder = null;
+        for (Map.Entry<Index, List<FailedRerouteAllocation.StaleShard>> entry : staleShards.stream().collect(
+            Collectors.groupingBy(fs -> fs.shardId.getIndex())).entrySet()) {
+            final IndexMetaData oldIndexMetaData = oldMetaData.getIndexSafe(entry.getKey());
```

Comment:
This is totally my ignorance, but is there a situation where an index could be removed but also have a shard marked as stale, in which case this would throw an exception because the index isn't in the metadata?